### DISCUSSION TITLE
Server/docker version

### DISCUSSION THREAD

**Discussion by Heroic-Brine** [mikulash/Octoprint-detector2#10] [Ideas]
Adding a version for servers would be nice such as a headless client, for both windows and Linux (although Linux is what i very much would like to see) a docker container is also a very simple way of making it compatible with more or less any system.

**Comment by mikulash** [comment]
doesn't the spaghetti detective uses a server for it's detection? my goal was to eliminate the need for any external device excepts the Raspberry and PC

**Comment by Heroic-Brine** [comment]
It dose, but yours seems to work better, although i have done very little testing.

While i love the idea you don't need a server it would be nice in my case as my computer is only so often turned on and my server is much more power efficient and is already on 24/7. my printer and octoprint are only online when i use them

**Comment by 1n5aN1aC** [comment]
Does the Pi4 have enough power to run the model?  I would think it would, and the detection could be ran directly on the octoprint host.

If not, how about only running it every x timelapse frames?  Only checking every 30 seconds for example, and running the detection process with a high nice level, so it doesn't impact printing.

**Comment by Heroic-Brine** [comment]
I doubt it or it would impact printing performance especially for people running multiple printers on one pi. Although I can test if you don't have a 4.

**Comment by 1n5aN1aC** [comment]
I think it looks like it's running a version of MobileNet, (not sure which version or size) which is a pretty small and performant neural network, made to do computation on mobile phones, so I really do think the Pi4 could handle it easily, probably even could handle a frame every second in theory.
Unsure about older Pis.

That being said, I haven't dug into the code too much, but since it's currently all browser-based, it could be harder to move it to run on the octoprint server itself, as you'd have to either port the plugin itself away from being primarily JavaScript, or have users install Node.js.

Anyways, I definitely think the net could run on the more modern Pis, and I think that would be by far the best solution, but I'm not sure how hard that would be to implement.....

**Reply by Heroic-Brine** [reply]
Might be  possible not sure if it is the best solution i would love to have it running on my own server every server

**Reply by mikulash** [reply]
The model was created with Python and later converted to tensorflowjs and I still have somewhere saved the original, so running it on the Pi would be theoretically possible. I think I could add an option to let the user choose, where the detection should be running.

**Reply by 1n5aN1aC** [reply]
That would be very cool!

I don't usually have a client connected to the webUI, usually just print via UI, then don't have it open for hours and hours, so if it could all be integrated into the octoprint plugin itself, and run on the server itself, that would amazingly impressive!

If you do decide to start work on that, let me know and I'd be glad to toss a couple coffees your way for your time!
